services/search/pkg/query/bleve/compiler.go

```diff
@@ -232,9 +242,21 @@ func mapBinary(operator *ast.OperatorNode, ln, rn bleveQuery.Query, leftIsGroup
 	}
 	if operator.Value == kql.BoolOR {
 		if left, ok := ln.(*bleveQuery.DisjunctionQuery); ok {
+			// if both are DisjunctionQuery then merge
+			if right, ok := rn.(*bleveQuery.DisjunctionQuery); ok {
+				left.AddQuery(right.Disjuncts...)
+				return left
+			}
 			left.AddQuery(rn)
 			return left
 		}
+		if _, ok := ln.(*bleveQuery.ConjunctionQuery); !ok {
+			if right, ok := rn.(*bleveQuery.DisjunctionQuery); ok {
+				left := bleveQuery.NewDisjunctionQuery([]bleveQuery.Query{ln})
+				left.AddQuery(right.Disjuncts...)
+				return left
+			}
+		}
```

Comment:
We are type asserting `ln` and `rn` multiple times here. Can't we solve that with a typeswitch? Like
```golang
right, ok := rn.(*bleveQuery.DisjunctionQuery)
switch left := ln.(type) {
default:
    if ok {
        left := bleveQuery.NewDisjunctionQuery([]bleveQuery.Query{ln})
        left.AddQuery(right.Disjuncts...)
        return left
    }
    fallthrough
case *bleveQuery.ConjunctionQuery:
    return bleveQuery.NewDisjunctionQuery([]bleveQuery.Query{ln, rn})
case *bleveQuery.DisjunctionQuery:
    if ok {
        left.AddQuery(right.Disjunkts)
    } else {
        left.AddQuery(rn)
   }
   return left	
}
```
Or similar. But we want to make the type assertions only once

Reply:
The fallthrough doesn't work in this way.
I can rewrite but It became badly readable, sounds to me
`		right, ok := rn.(*bleveQuery.DisjunctionQuery)
		switch left := ln.(type) {
		case *bleveQuery.DisjunctionQuery:
			if ok {
				left.AddQuery(right.Disjuncts...)
			} else {
				left.AddQuery(rn)
			}
			return left
		case *bleveQuery.ConjunctionQuery:
			return bleveQuery.NewDisjunctionQuery([]bleveQuery.Query{ln, rn})
		default:
			if ok {
				left := bleveQuery.NewDisjunctionQuery([]bleveQuery.Query{ln})
				left.AddQuery(right.Disjuncts...)
				return left
			} else {
				return bleveQuery.NewDisjunctionQuery([]bleveQuery.Query{ln, rn})
			}
		}`

Reply:
Imo better readable than the double type assertion. If you ask me I'd go with second solution. But up to you in the end...

Reply:
reworked